src/elfpy/markets.py

```diff
@@ -176,38 +215,29 @@ def trade_and_update(self, agent_action):
             LP tokens are also stored in user wallet as fungible amounts, for ease of use
         """
         self.check_action_type(agent_action.action_type)
+
         # TODO: check the desired amount is feasible, otherwise return descriptive error
         # update market variables which may have changed since the user action was created
-        agent_action.time_remaining = time_utils.get_yearfrac_remaining(
-            self.time, agent_action.mint_time, self.token_duration
-        )
-        agent_action.stretched_time_remaining = time_utils.stretch_time(
-            agent_action.time_remaining, self.time_stretch_constant
-        )
+        time_remaining = time_utils.get_yearfrac_remaining(self.time, agent_action.mint_time, self.token_duration)
+        stretched_time_remaining = time_utils.stretch_time(time_remaining, self.time_stretch_constant)
         # if self.verbose:
         agent_action.print_description_string()
         # for each position, specify how to forumulate trade and then execute
+        # TODO: we shouldn't set values on the object passed in, add parameters to
+        # open/close_long/short functions so that this isn't necessary
         update_price_and_rate = True
         if agent_action.action_type == "open_long":  # buy to open long
-            agent_action.direction = "out"  # calcOutGivenIn
-            agent_action.token_out = "pt"  # sell known base for unknown PT
-            market_deltas, agent_deltas = self._open_long(agent_action)
+            market_deltas, agent_deltas = self._open_long(agent_action, "pt", stretched_time_remaining)
         elif agent_action.action_type == "close_long":  # sell to close long
-            agent_action.direction = "out"  # calcOutGivenIn
-            agent_action.token_out = "base"  # sell known PT for unknown base
-            market_deltas, agent_deltas = self._close_long(agent_action)
+            market_deltas, agent_deltas = self._close_long(agent_action, "base", stretched_time_remaining)
         elif agent_action.action_type == "open_short":  # sell PT to open short
-            agent_action.direction = "in"  # calcOutGivenIn
-            agent_action.token_out = "pt"  # sell known PT for unknown base
-            market_deltas, agent_deltas = self._open_short(agent_action)
+            market_deltas, agent_deltas = self._open_short(agent_action, "pt", time_remaining)
```

Comment:
I think open_short wants `stretched_time_remaining` as an argument to `calc_out_given_in`

Reply:
really good catch, didn't know this.